File: 799-minimum-distance-between-bst-nodes/minimum-distance-between-bst-nodes.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def __init__(self):
        self.prev=None
        self.min_diff=10**5+1

    def fun(self,root):
        if root!=None:
            self.fun(root.left)

            if self.prev==None:
                self.prev=root
            else:
                self.min_diff=min(self.min_diff,root.val-self.prev.val)
                self.prev=root
            self.fun(root.right)
    def minDiffInBST(self, root: Optional[TreeNode]) -> int:
        self.fun(root)
        return self.min_diff

File: 799-minimum-distance-between-bst-nodes/minimum-distance-between-bst-nodes.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    TreeNode prev = null;
     int minDiff = Integer.MAX_VALUE;

    void minDiff(TreeNode root) {
        if(root!=null) {
            minDiff(root.left);
            if(prev == null) prev = root;
            else {
                minDiff = Math.min(minDiff, (root.val - prev.val));
                System.out.print(root.val+" ");
                System.out.print(prev.val+" ");
                System.out.print(minDiff+" ");
                prev = root;
                
            }
            minDiff(root.right);
        }
        return;
    }
    public int minDiffInBST(TreeNode root) {
        minDiff(root);
        return minDiff;
    }
}